### PR TITLE
Remove support for "dims on the fly"

### LINKED ISSUE #6085: Remove support of "dims on the fly"
### Discussed in https://github.com/pymc-devs/pymc/discussions/6065

<div type='discussions-op-text'>

<sup>Originally posted by **ricardoV94** August 25, 2022</sup>
Right now we can specify dims on the fly, which are taken from the shape of a variable:

```python
import pymc as pm
with pm.Model() as m:
  x = pm.Normal("x", shape=(3,), dims=("trial",))
  print(m.dim_lengths["trial"].eval())  # 3
```

This is neat, but causes some code complexity, because we also allow dims to inform the shape of a variable.
```python
with pm.Model(coords={"trial": range(3)}) as m:
  x = pm.Normal("x", dims=("trial",))
  print(x.eval().shape)  # (3,)
```

Which forces us to create new variables in a two-step process, first without dim information and then with it (via resizing). Code could be simplified if we don't have to worry about dims being specified on the fly. 

These two behaviors of dims are also potentially confusing when mixed together:
```python
with pm.Model() as m:
  x = pm.Normal("x", shape=5, dims=("trial",))
  y = pm.Normal("y", dims=("trial",))
  print(y.eval().shape)  # (5,)
```
</div>

### PR DESCRIPTION
**Pull request by ricardoV94**
Closes #6085

**Checklist**
+ [x] Explain important implementation details 👆
+ [x] Make sure that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html).
+ [x] Link relevant issues (preferably in [nice commit messages](https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html))
+ [x] Are the changes covered by tests and docstrings?
+ [x] Fill out the short summary sections 👇

## Major / Breaking Changes
- Dims can no longer be implied by the size of a variable or observed, they must always be defined in advance

## Bugfixes / New features
- ...

## Docs / Maintenance
- ...

### CODE REVIEW

**Review by twiecki — APPROVED** [review]

**Comment by ricardoV94** [comment]
Marking as a draft because some checks we had in `model.set_data` may no longer make sense?

**Comment by ricardoV94** [comment]
Failing test is unrelated: #6091

File: pymc/model.py [hunk 1]
```diff
@@ -1295,6 +1295,7 @@ def set_data(
                         )
                     else:
                         # The dimension is immutable.
+                        # TODO: Is this still possible?
```
Comment (ricardoV94) [inline pymc/model.py]: @michaelosthege  can I get your input here? Do you remember if this condition related only to inferred dims?

Comment (michaelosthege) [reply pymc/model.py]: This condition is related to dims that were defined by the `shape` of a non-shared `TensorVariable`.

This was the case for inferred dims, but there are at least unintended ways how to end up here:

```python
with pm.Model() as pmodel:
    x = pm.Normal("x", size=7)

    pmodel.dim_lengths["weekday"] = x.shape[0]       # this definitely
    pmodel.add_coord("weekday", length=x.shape[0])   # maybe also this

    pm.MutableData("y", np.arange(7), dims="weekday")
pmodel.set_data("y", np.arange(10))
```

Comment (ricardoV94) [reply pymc/model.py]: Isn't this another form of "dims on the fly"?

Comment (ricardoV94) [reply pymc/model.py]: Anyway added your example to cover that branch

**Pull request merged**